src/inmanta/deploy/state.py

```diff
@@ -234,38 +199,60 @@ def block_provides(self, resources: Set[ResourceIdStr]) -> set[ResourceIdStr]:
         :param resources: The set of resources for which the provides have to be blocked.
         :return: The set of dependent resources that were marked as blocked (transitively).
         """
-        result = set()
-        for dependent_resource in self.requires.get_all_provides_transitively(resources):
-            self.block_resource(dependent_resource, self.resources[dependent_resource], transient=True)
-            result.add(dependent_resource)
+        result: set[ResourceIdStr] = set()
+        provides_view: Mapping[ResourceIdStr, Set[ResourceIdStr]] = self.requires.provides_view()
+        todo: list[ResourceIdStr] = list(itertools.chain.from_iterable(provides_view.get(r, set()) for r in resources))
+        while todo:
+            resource_id: ResourceIdStr = todo.pop()
+            if self.resource_state[resource_id].blocked.is_blocked():
+                # Resource is already blocked. All provides will be blocked as well.
+                continue
+            else:
+                # Block resource and its provides.
+                self.block_resource(resource_id, self.resources[resource_id], is_transitive=True)
+                result.add(resource_id)
+                todo.extend(provides_view.get(resource_id, set()))
         return result
 
-    def unblock_resource(self, resource: ResourceIdStr) -> None:
+    def mark_as_defined(self, resource: ResourceIdStr, details: ResourceDetails) -> None:
         """
-        Mark the given resource no longer as blocked. Also mark the provides of the given resource as unblocked (transitively)
-        if they become unblocked because of this.
+        Mark the given resource as defined. If the given resource or any of its provides became unblocked, because of this,
+        this method also updates the blocked status.
 
         Must be called under the scheduler lock. This method assumes that all the resources of the model version
         are populated in the resources dictionary.
 
-        :param resource: The resource that has to be unblocked and potentially any of its provides as a consequence of
-                         `resource` being unblocked.
+        :param resource: The resource that became defined.
+        :param details: The details of resource.
         """
 
-        def _unblock(resource: ResourceIdStr) -> None:
-            self.resource_state[resource].blocked = BlockedStatus.NO
+        def update_blocked_status(resource: ResourceIdStr) -> bool:
+            """
+            Check if the deployment of the given resource is still blocked and mark it as unblocked if it is.
+
+            :return: True iff the given resource was unblocked.
+            """
             if self.resource_state[resource].status is ResourceStatus.UNDEFINED:
```

Comment:
We should probably also have a check for `self.resource_state[resource].blocked is BlockedStatus.NO` (or keep track of these in a set) to make sure we don't go down the same subtree twice in the case of diamond provides.